From 575a4bd9a5774104805aefd584112dbf3ceba3f8 Mon Sep 17 00:00:00 2001
From: Franck LENORMAND <franck.lenormand@nxp.com>
Date: Thu, 21 Jun 2018 15:20:14 +0200
Subject: [PATCH 4204/5242] MLK-18082: crypto: caam: sm: Fix descriptor
 running functions

commit  ddfe4a0db32d41ab2bcf05f9af0214e5a5d03c97 from
https://source.codeaurora.org/external/imx/linux-imx.git

In case of error when runnign descriptor, there was no indication
of the root cause with the appropriate existing function.

Signed-off-by: Franck LENORMAND <franck.lenormand@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/crypto/caam/sm_store.c |   17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/drivers/crypto/caam/sm_store.c b/drivers/crypto/caam/sm_store.c
index a342756..df18f1b 100644
--- a/drivers/crypto/caam/sm_store.c
+++ b/drivers/crypto/caam/sm_store.c
@@ -420,13 +420,17 @@ void sm_key_job_done(struct device *dev, u32 *desc, u32 err, void *context)
 {
 	struct sm_key_job_result *res = context;
 
+	if (err)
+		caam_jr_strstatus(dev, err);
+
 	res->error = err;	/* save off the error for postprocessing */
+
 	complete(&res->completion);	/* mark us complete */
 }
 
 static int sm_key_job(struct device *ksdev, u32 *jobdesc)
 {
-	struct sm_key_job_result testres;
+	struct sm_key_job_result testres = {0};
 	struct caam_drv_private_sm *kspriv;
 	int rtn = 0;
 
@@ -436,10 +440,13 @@ static int sm_key_job(struct device *ksdev, u32 *jobdesc)
 
 	rtn = caam_jr_enqueue(kspriv->smringdev, jobdesc, sm_key_job_done,
 			      &testres);
-	if (!rtn) {
-		wait_for_completion_interruptible(&testres.completion);
-		rtn = testres.error;
-	}
+	if (rtn)
+		goto exit;
+
+	wait_for_completion_interruptible(&testres.completion);
+	rtn = testres.error;
+
+exit:
 	return rtn;
 }
 
-- 
1.7.9.5

